File: 74.cpp
int OneB(const vector<vector<int>>& matrix, int index) {
	return matrix[index / matrix[0].size()][index % matrix[0].size()];
}

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
    	int l = 0, r = matrix.size() * matrix[0].size();
    	while(l <= r) {
    		int mid = (l + r) / 2;
    		if(OneB(matrix, mid) == target) return true;
    		else if(OneB(matrix, mid) > target) r = mid - 1;
    		else l = mid + 1;
    	}
    	return false;
    }
};

File: 74.py
from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        first_col = [matrix[i][0] for i in range(m)]
        i = bisect.bisect(first_col, target) - 1
        if i > 0 and first_col[i] == target:
            return True
        j = bisect.bisect_left(matrix[i], target)
        return j < n and matrix[i][j] == target
